lib/spack/spack/util/windows_registry.py

```diff
@@ -59,49 +59,111 @@ def subkeys(self):
     def hkey(self):
         return self._handle
 
+    def _handle_winreg_errors(self, err, name, *args, **kwargs):
+        # Expected errors that occur on occasion, these are easily
+        # debug-able and have sufficiently verbose reporting and obvious cause
+        # [WinError 2]: the system cannot find the file specified - lookup item does
+        # not exist
+        # [WinError 5]: Access is denied - user does not have access to key
+        if hasattr(err, "winerror") and err.winerror in (5, 2):
+            raise err
+        # Other OS errors are more difficult to diagnose, so we wrap them in some extra
+        # reporting
+        raise InvalidRegistryOperation(name, err, self.path, *args, **kwargs) from err
+
+    def OpenKeyEx(self, subname, **kwargs):
+        """Convenience wrapper around winreg.OpenKeyEx"""
+        tty.debug(
+            f"[WINREG ACCESS] Accessing Reg Key {self.path}/{subname} with"
+            f" {kwargs.get('access', 'default')} access"
+        )
+        try:
+            return winreg.OpenKeyEx(self.hkey, subname, **kwargs)
+        except OSError as e:
+            self._handle_winreg_errors(e, "OpenKeyEx", subname, **kwargs)
+
+    def QueryInfoKey(self):
+        """Convenience wrapper around winreg.QueryInfoKey"""
+        tty.debug(f"[WINREG ACCESS] Obtaining key,value information from key {self.path}")
+        try:
+            return winreg.QueryInfoKey(self.hkey)
+        except OSError as e:
+            self._handle_winreg_errors(e, "QueryInfoKey")
+
+    def EnumKey(self, index):
+        """Convenience wrapper around winreg.EnumKey"""
+        tty.debug(
+            "[WINREG ACCESS] Obtaining name of subkey at index "
+            f"{index} from registry key {self.path}"
+        )
+        try:
+            return winreg.EnumKey(self.hkey, index)
+        except OSError as e:
+            self._handle_winreg_errors(e, "EnumKey", index)
+
+    def EnumValue(self, index):
+        """Convenience wrapper around winreg.EnumValue"""
+        tty.debug(
+            f"[WINREG ACCESS] Obtaining value at index {index} from registry key {self.path}"
+        )
+        try:
+            return winreg.EnumValue(self.hkey, index)
+        except OSError as e:
+            self._handle_winreg_errors(e, "EnumValue", index)
+
+    def QueryValueEx(self, name, **kwargs):
+        """Convenience wrapper around winreg.QueryValueEx"""
+        tty.debug(f"[WINREG ACCESS] Obtaining value {name} from registry key {self.path}")
+        try:
+            return winreg.QueryValueEx(self.hkey, name, **kwargs)
+        except OSError as e:
+            # Expected errors that occur on occasion, these are easily
+            # debug-able and have sufficiently verbose reporting and obvious cause
+            if hasattr(e, "winerror") and e.winerror in (5, 2):
+                raise e
```

Comment:
A couple things:

* this appears to duplicate logic from `_handle_winreg_errors`
* some repetition could be avoided by defining a wrapper/context-manager for

```
        try:
            ...
        except OSError as e:
            self._handle_winreg_errors(e, "EnumKey", index)
```

like

```
with self.error_handler():
  return winreg.QueryValueEx(self.hkey, name, **kwargs)
```

Reply:
Great suggestion, will implement.